src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -163,5 +176,107 @@ private IEnumerable<CodeRefactoringProvider> GetProviders(Document document)
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private class ProjectCodeRefactoringProvider
+        {
+            private readonly AnalyzerReference _reference;
+            private ImmutableDictionary<string, ImmutableArray<CodeRefactoringProvider>> _refactoringsPerLanguage;
+
+            public ProjectCodeRefactoringProvider(AnalyzerReference reference)
+            {
+                _reference = reference;
+                _refactoringsPerLanguage = ImmutableDictionary<string, ImmutableArray<CodeRefactoringProvider>>.Empty;
+            }
+
+            public ImmutableArray<CodeRefactoringProvider> GetRefactorings(string language)
+            {
+                return ImmutableInterlocked.GetOrAdd(ref _refactoringsPerLanguage, language, CreateRefactorings);
+            }
+
+            private ImmutableArray<CodeRefactoringProvider> CreateRefactorings(string language)
+            {
+                // check whether the analyzer reference knows how to return fixers directly.
+                if (_reference is ICodeRefactoringProviderFactory codeRefactoringProviderFactory)
+                {
+                    return codeRefactoringProviderFactory.GetRefactorings();
+                }
+
+                // otherwise, see whether we can pick it up from reference itself
+                var analyzerFileReference = _reference as AnalyzerFileReference;
+                if (analyzerFileReference == null)
+                {
+                    return ImmutableArray<CodeRefactoringProvider>.Empty;
+                }
+
+                IEnumerable<TypeInfo> typeInfos = null;
+                var builder = ArrayBuilder<CodeRefactoringProvider>.GetInstance();
+
+                try
+                {
+                    Assembly analyzerAssembly = analyzerFileReference.GetAssembly();
+                    typeInfos = analyzerAssembly.DefinedTypes;
+
+                    foreach (var typeInfo in typeInfos)
+                    {
+                        if (typeInfo.IsSubclassOf(typeof(CodeRefactoringProvider)))
+                        {
+                            try
+                            {
+                                var attribute = typeInfo.GetCustomAttribute<ExportCodeRefactoringProviderAttribute>();
+                                if (attribute != null)
+                                {
+                                    if (attribute.Languages == null ||
+                                        attribute.Languages.Length == 0 ||
+                                        attribute.Languages.Contains(language))
+                                    {
+                                        builder.Add((CodeRefactoringProvider)Activator.CreateInstance(typeInfo.AsType()));
+                                    }
+                                }
+                            }
+                            catch
+                            {
+                            }
```

Comment:
pity we can't report anything.  do we have any sort of general service that might even just take a string and put it in the output window.  i could imagine people tearing their hair out going "why isn't my provider loading in roslyn!?!" (and it turns out the constructor is inaccessible or something).

Reply:
@mavasani Has this been a problem for code fixes?

Reply:
I don't recall any customer feedback on this, but very likely we have some tracking work item to make such failures more discoverable. @heejaechang do you recall?

Reply:
> I don't recall any customer feedback on this,

note: i've personally run into it.  it's def a pita. but it's also something that isn't really a deal breaker.  it would just be nice to fixup *if* we're currently working in this area :)

Reply:
Leaving this for now for consistency against existing services.

Reply:
Can you move this type to a separate file?